### ISSUE TITLE
Causes issue with Access-Control-Allow-Origin:*

### ISSUE THREAD

**Issue by danielchatfield** [EFForg/https-everywhere#49]
Browser: chrome
Example (~~may be fixed soon though~~ Fixed by using https directly): http://yeoman.io/community-generators.html

Page tries to load http://yeoman-generator-list.herokuapp.com/ which results in: `XMLHttpRequest cannot load http://yeoman-generator-list.herokuapp.com/. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://yeoman.io' is therefore not allowed access.`.

If the page requests the https version directly then it works.

**Comment by diracdeltas** [comment]
anne van kesteren says this Firefox bug might be relevant: https://bugzilla.mozilla.org/show_bug.cgi?id=881830

**Comment by diracdeltas** [comment]
Bug in chrome: https://code.google.com/p/chromium/issues/detail?id=387198

**Comment by diracdeltas** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=881830#c12 seems like the right explanation here. When a subresource is redirected to HTTPS by HTTPS-E or HSTS, the browser says, "I have no CORS information for this new URL, so just block it."

Did a quick test and this seems to be fixed in Chrome but not in Firefox. To repro, make sure you have the Heroku rule enabled in the dev version of HTTPS Everywhere Firefox, then do an XHR to `http://yeoman-generator-list.herokuapp.com/` in the context of an HTTP page at a different origin. In Firefox 36, the XHR silently fails. OTOH, `https://yeoman-generator-list.herokuapp.com/` loads fine.

**Comment by SimonSapin** [comment]
That Firefox bug is [marked as fixed in Firefox 38](https://bugzilla.mozilla.org/show_bug.cgi?id=881830#c40), but I can reproduce #1365 (which I think is a case of this bug) on Firefox Nightly 40.0a1 (2015-04-04)

**Comment by AnAkIn1** [comment]
I still have the same issue with Firefox 39.

**Comment by AnAkIn1** [comment]
Is there any progress on this? I can't enable https everywhere because I have this issue on one of the websites I often visit.

**Comment by jsha** [comment]
No progress so far, I'm afraid. In the meantime, if you would like to use HTTPS Everywhere, you can disable the rules for the specific site that has problems using the icon menu.

**Comment by TheNavigat** [comment]
https://github.com/EFForg/https-everywhere/issues/2637

**Comment by TheNavigat** [comment]
Just reminding everyone that the issue still exists.

This might be the closest I got to reproducing it. http://origin-static.gamespot.com/js/modals.json

Using Chrome inspector I found that the response headers are different, depending on whether the extension is enabled or not.

These are the headers with the extension enabled.

HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Vary: Accept-Encoding
X-Powered-By: PHP/5.4.17
Cache-Control: no-cache, private
Server: nginx/1.2.0
Vary: Accept-Encoding
Date: Sat, 22 Aug 2015 15:17:00 GMT
X-TX-ID: 882debef-d7c8-4efa-85ad-f407b05eaa52
X-Phoenix-Id: 55d8926bd79fd
X-Varnish: 35127409
Age: 0
Via: 1.1 varnish-v4
X-Cache-Debug-Served-By: ws420.phx2.cbsig.net
X-Cache-Debug-Result: miss
X-Cache-Device: other
X-Cache-Region: us-uk
X-Cache-AB: 38-0|42-0|44-0|45-0|47-0|50-0|52-0|53-0
Set-Cookie: vguid=f69bf9fb-572d-4660-9757-62243fa7d213; path=/
Content-Encoding: gzip

And these are the headers with the extension disabled.

HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Vary: Accept-Encoding
X-Powered-By: PHP/5.4.17
Server: nginx/1.2.0
Vary: Accept-Encoding
X-TX-ID: 76e69086-e19d-407e-b83d-04b83136f8b8
X-Phoenix-Id: 55d891ad1146e
X-Varnish: 33161980
X-Cache-Debug-Served-By: ws420.phx2.cbsig.net
X-Cache-Debug-Result: miss
X-Cache-Device: other
X-Cache-Region: us-uk
X-Cache-AB: 38-0|42-0|44-0|45-0|47-0|50-0|52-0|53-0
Content-Encoding: gzip
Cache-Control: private, max-age=2592000
Date: Sat, 22 Aug 2015 15:13:49 GMT
Content-Length: 1357
Connection: keep-alive
Set-Cookie: vguid=6283bcd1-d1c1-4ee6-b404-65e49d14a1bb; path=/
access-control-allow-origin: *

**Comment by TheNavigat** [comment]
Is creating a function that checks whether an Access-Control-Allow-Origin header exists and then injects it after redirection a good idea?

**Comment by jsha** [comment]
That's an interesting idea but somewhat worrying: There is a chance we could create a security vulnerability on some websites if we implement it wrong.

**Comment by TheNavigat** [comment]
Maybe we can specify which domains to add the header to, then, like what we're doing with the ruleset for example.

**Comment by berrythesoftwarecodeprogrammar** [comment]
i think this is the closest issue i could find to mine. CORS requests recently have been failing on every single site. only fixed by disabling https-everywhere. not sure why

**Comment by swrobel** [comment]
I get the same issue on http://js2.coffee using Firefox 43.0.2 & HTTPS-Everywhere 5.1.2 on a Mac. I do not have the same issue on Chrome 47 & HTTPS-Everywhere 2015.12.16. The Firefox console shows the following error:

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://code.ionicframework.com/ionicons/2.0.1/fonts/ionicons.woff?v=2.0.1. (Reason: CORS header 'Access-Control-Allow-Origin' missing).

Checking the headers of both http://code.ionicframework.com/ionicons/2.0.1/fonts/ionicons.woff?v=2.0.1 and https://code.ionicframework.com/ionicons/2.0.1/fonts/ionicons.woff?v=2.0.1 reveals that they include `Access-Control-Allow-Origin:*` as expected.

**Comment by chojrak11** [comment]
I'm bumping this with problem on www.easyjet.com. The button "Check-in Online" doesn't work when HTTPS Everywhere is installed/enabled in Firefox. Version 46.0beta8, HTTPS Everwhere 5.1.6. The error message in Firefox console is:

> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://www.easyjet.com/ejcms/nocache/jsCallbacks/CheckInCallback.aspx?qs=&scLang=pl-PL&sender=. (Reason: CORS header Access-Control-Allow-Origin missing)

**Comment by mbergin** [comment]
Seeing this in 2016.9.1 in Chrome 52.0.2743.116 which breaks the flight search form:

XMLHttpRequest cannot load https://www.easyjet.com/ejcms/cache15m/api/flights/search?AdminFeePerBookin…,NTE,NCE,TLS,AJA,BIA,FSC,CDG,ORY,SXF&MaxResults=4&PreferredOriginIatas=LGW. Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials. Origin 'http://www.easyjet.com' is therefore not allowed access.

**Comment by mayhemer** [comment]
Regarding Firefox, I filed a Mozilla bug to add an API to support HTTP->HTTPS redirects, that would not be blocked by CORS checks, invoked by an XPCOM API directly on http channels.  I never got a feedback from Yan in our bugzilla, tho.

Bumping here then, since the issue still exists.  Please see https://bugzilla.mozilla.org/show_bug.cgi?id=1149250#c3 for suggested changes in Gecko and give a feedback here or there please.  We want to move forward.  Thanks.

**Comment by herrernst** [comment]
This is now fixed in upcoming Firefox 59 with "upgradeToSecure", see https://bugzilla.mozilla.org/show_bug.cgi?id=1149250

Would be good if HTTPS Everywhere would use it in the future to reduce breakage

**Comment by jeremyn** [comment]
It seems like from https://bugzilla.mozilla.org/show_bug.cgi?id=1149250 and https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/webRequest/BlockingResponse that there is a new option in Firefox that allows some `http`-to-`https` rewrites to bypass CORS checks, which would let us avoid a lot of errors. This needs to be implemented though, and it doesn't seem to be an option yet in Chrome according to https://developer.chrome.com/extensions/webRequest.

I think this is high priority, or at least, something I don't want to get lost. How should we track progress on implementing this? Should we make a new issue?

Ping @Bisaloo @brainwane @gloomy-ghost @Hainish @J0WI @wonderchook .

**Comment by gloomy-ghost** [comment]
> it doesn't seem to be an option yet in Chrome 

IIRC Chrome managed to fix the issue without forcing extensions to use a new option, so it is only broken with Firefox.

**Comment by jeremyn** [comment]
@gloomy-ghost That seems to be true, for example, from issue https://github.com/EFForg/https-everywhere/issues/14117, http://shop.decatorevista.ro/ has working styling and no CORS console messages in Chrome, but has broken styling with CORS console messages in Firefox. I don't know what Chrome does or doesn't specifically allow, though.
